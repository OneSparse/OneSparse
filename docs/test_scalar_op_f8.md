# Scalar `float8` Operations

Test various scalar math operations with native Postgres types
``` postgres-console
select (1::float8)::scalar + -1::float8;
  ?column?   
-------------
 f8:0.000000
(1 row)

select 1::float8 + (-1::float8)::scalar;
 ?column? 
----------
        0
(1 row)

select (1::float8)::scalar - 1::float8;
  ?column?   
-------------
 f8:0.000000
(1 row)

select 1::float8 - (1::float8)::scalar;
 ?column? 
----------
        0
(1 row)

select (1::float8)::scalar * -1::float8;
   ?column?   
--------------
 f8:-1.000000
(1 row)

select 1::float8 * (-1::float8)::scalar;
 ?column? 
----------
       -1
(1 row)

select (1::float8)::scalar / 1::float8;
  ?column?   
-------------
 f8:1.000000
(1 row)

select 1::float8 / (1::float8)::scalar;
 ?column? 
----------
        1
(1 row)

```
Test construction of min, zero and max values:
``` postgres-console
select '-1'::float8::scalar;
    scalar    
--------------
 f8:-1.000000
(1 row)

select '0'::float8::scalar;
   scalar    
-------------
 f8:0.000000
(1 row)

select '1'::float8::scalar;
   scalar    
-------------
 f8:1.000000
(1 row)

```
Test setting a scalar value from max to 2
``` postgres-console
select set('1'::float8::scalar, 2);
     set     
-------------
 f8:2.000000
(1 row)

```
Test various casting functions used by the CREATE CAST machinery:
``` postgres-console
select scalar_float8((-1)::float8);
 scalar_float8 
---------------
 f8:-1.000000
(1 row)

select scalar_float8((0)::float8);
 scalar_float8 
---------------
 f8:0.000000
(1 row)

select scalar_float8((1)::float8);
 scalar_float8 
---------------
 f8:1.000000
(1 row)

select float8_scalar((-1)::float8::scalar);
 float8_scalar 
---------------
            -1
(1 row)

select float8_scalar((0)::float8::scalar);
 float8_scalar 
---------------
             0
(1 row)

select float8_scalar((1)::float8::scalar);
 float8_scalar 
---------------
             1
(1 row)

```
These casting functions cast the Postgres type `float8` to the
GraphBLAS scalar type `GrB_INT64`.
``` postgres-console
select cast(-1::float8 as scalar);
    scalar    
--------------
 f8:-1.000000
(1 row)

select cast(0::float8 as scalar);
   scalar    
-------------
 f8:0.000000
(1 row)

select cast(1::float8 as scalar);
   scalar    
-------------
 f8:1.000000
(1 row)

```
These tests cast back from the scalar type `GrB_INT64` to the
Postgres type `float8`
``` postgres-console
select cast((-1::float8)::scalar as float8);
 float8 
--------
     -1
(1 row)

select cast((0::float8)::scalar as  float8);
 float8 
--------
      0
(1 row)

select cast((1::float8)::scalar as scalar);
   scalar    
-------------
 f8:1.000000
(1 row)

```