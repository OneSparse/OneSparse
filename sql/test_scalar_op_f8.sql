-- # Scalar `float8` Operations
--
-- Test various scalar math operations with native Postgres types
select (1::float8)::scalar + -1::float8;
select 1::float8 + (-1::float8)::scalar;
select (1::float8)::scalar - 1::float8;
select 1::float8 - (1::float8)::scalar;
select (1::float8)::scalar * -1::float8;
select 1::float8 * (-1::float8)::scalar;
select (1::float8)::scalar / 1::float8;
select 1::float8 / (1::float8)::scalar;

-- Test construction of min, zero and max values:
select '-1'::float8::scalar;
select '0'::float8::scalar;
select '1'::float8::scalar;

-- Test setting a scalar value from max to 2
select set('1'::float8::scalar, 2);

-- Test various casting functions used by the CREATE CAST machinery:
select scalar_float8((-1)::float8);
select scalar_float8((0)::float8);
select scalar_float8((1)::float8);
select float8_scalar((-1)::float8::scalar);
select float8_scalar((0)::float8::scalar);
select float8_scalar((1)::float8::scalar);

-- These casting functions cast the Postgres type `float8` to the
-- GraphBLAS scalar type `GrB_INT64`.

select cast(-1::float8 as scalar);
select cast(0::float8 as scalar);
select cast(1::float8 as scalar);

-- These tests cast back from the scalar type `GrB_INT64` to the
-- Postgres type `float8`

select cast((-1::float8)::scalar as float8);
select cast((0::float8)::scalar as  float8);
select cast((1::float8)::scalar as scalar);
