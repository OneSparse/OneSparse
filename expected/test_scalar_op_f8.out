-- # Scalar `float8` Operations
--
-- Test various scalar math operations with native Postgres types
select (1::float8)::scalar + -1::float8;
   ?column?    
---------------
 fp64:0.000000
(1 row)

select 1::float8 + (-1::float8)::scalar;
 ?column? 
----------
        0
(1 row)

select (1::float8)::scalar - 1::float8;
   ?column?    
---------------
 fp64:0.000000
(1 row)

select 1::float8 - (1::float8)::scalar;
 ?column? 
----------
        0
(1 row)

select (1::float8)::scalar * -1::float8;
    ?column?    
----------------
 fp64:-1.000000
(1 row)

select 1::float8 * (-1::float8)::scalar;
 ?column? 
----------
       -1
(1 row)

select (1::float8)::scalar / 1::float8;
   ?column?    
---------------
 fp64:1.000000
(1 row)

select 1::float8 / (1::float8)::scalar;
 ?column? 
----------
        1
(1 row)

-- Test construction of min, zero and max values:
select '-1'::float8::scalar;
     scalar     
----------------
 fp64:-1.000000
(1 row)

select '0'::float8::scalar;
    scalar     
---------------
 fp64:0.000000
(1 row)

select '1'::float8::scalar;
    scalar     
---------------
 fp64:1.000000
(1 row)

-- Test setting a scalar value from max to 2
select set('1'::float8::scalar, 2);
      set      
---------------
 fp64:2.000000
(1 row)

-- Test various casting functions used by the CREATE CAST machinery:
select scalar_float8((-1)::float8);
 scalar_float8  
----------------
 fp64:-1.000000
(1 row)

select scalar_float8((0)::float8);
 scalar_float8 
---------------
 fp64:0.000000
(1 row)

select scalar_float8((1)::float8);
 scalar_float8 
---------------
 fp64:1.000000
(1 row)

select float8_scalar((-1)::float8::scalar);
 float8_scalar 
---------------
            -1
(1 row)

select float8_scalar((0)::float8::scalar);
 float8_scalar 
---------------
             0
(1 row)

select float8_scalar((1)::float8::scalar);
 float8_scalar 
---------------
             1
(1 row)

-- These casting functions cast the Postgres type `float8` to the
-- GraphBLAS scalar type `GrB_INT64`.
select cast(-1::float8 as scalar);
     scalar     
----------------
 fp64:-1.000000
(1 row)

select cast(0::float8 as scalar);
    scalar     
---------------
 fp64:0.000000
(1 row)

select cast(1::float8 as scalar);
    scalar     
---------------
 fp64:1.000000
(1 row)

-- These tests cast back from the scalar type `GrB_INT64` to the
-- Postgres type `float8`
select cast((-1::float8)::scalar as float8);
 float8 
--------
     -1
(1 row)

select cast((0::float8)::scalar as  float8);
 float8 
--------
      0
(1 row)

select cast((1::float8)::scalar as scalar);
    scalar     
---------------
 fp64:1.000000
(1 row)

